test/reduce.cpp

```diff
@@ -685,3 +685,22 @@ TEST(Reduce, AllSmallIndexed) {
     array b       = a(seq(len / 2), span);
     ASSERT_EQ(max<float>(b), len / 2 - 1);
 }
+
+TEST(ProductAll, BoolIn_ISSUE2543_All_Ones) {
+    ASSERT_EQ(true, product<int>(constant(1, 5, 5, b8)) > 0);
```

Comment:
Shouldn't product return char?

Wouldn't the Result be one? In that case use

```

```
```suggestion
    ASSERT_EQ(1, product<char>(constant(1, 5, 5, b8));
```